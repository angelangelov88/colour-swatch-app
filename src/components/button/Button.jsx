import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'


const Button = (props) => {
  let {
    children,
    variant,
    size,
    rounded,
    tabIndex,
    type,
    disabled,
    onClick,
    className,
    ariaLabel
  } = props

  const classes = classNames(
    "font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center border",
    "focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary-light",
    variant === "primary" && "disabled:text-primary-10 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-30 bg-primary-light shadow disabled:border-primary-20 border-primary-dark",
    variant === "secondary" && "disabled:text-gray-50 active:text-gray-120 text-gray-90 hover:bg-gray-10 active:bg-gray-20 disabled:bg-gray-10 bg-white shadow border-gray-50",
    variant === "tertiary" && "disabled:text-primary-20 text-primary-light hover:bg-primary-10 active:bg-primary-20 disabled:bg-white bg-white disabled:border-primary-20 border-primary-10",
    size === 'base' && ``,
    size === 'sm' && `text-sm py-2 px-4`,
    size === 'md' && `text-lg py-2.5 px-5`,
    size === 'lg' && `text-xl py-3 px-6`,
    size === 'icon' && `p-2`,
    rounded === 'none' && `rounded-none`,
    rounded === 'sm' && `rounded-sm`,
    rounded === 'md' && `rounded-md`,
    rounded === 'lg' && `rounded-lg`,
    rounded === 'full' && `rounded-full p-2`,
    disabled && "disabled:cursor-default",
    className
  );

  return (
    <button
      type={type}
      disabled={disabled}
      onClick={onClick}
      aria-label={ariaLabel}
      tabIndex={tabIndex}
      className={classes}
    >
      {children}
    </button>
  )
}

Button.defaultProps = {
  children: undefined,
  variant: "primary",
  size: "sm",
  rounded: "full",
  type: "button",
  disabled: false,
  tabIndex: undefined,
  onClick: undefined,
  className: undefined,
  ariaLabel: undefined
}

Button.propTypes = {
  children: PropTypes.node,
  variant: PropTypes.oneOf(["primary", "secondary", "tertiary"]),
  size: PropTypes.oneOf(["base", "sm", "md", "lg", "icon"]), // @todo base
  rounded: PropTypes.oneOf(["none", "sm", "md", "lg", "full"]),
  type: PropTypes.oneOf(["button", "submit", "reset"]),
  disabled: PropTypes.bool,
  tabIndex: PropTypes.string,
  onClick: PropTypes.func,
  className: PropTypes.string,
  ariaLabel: PropTypes.string
}

export { Button }
