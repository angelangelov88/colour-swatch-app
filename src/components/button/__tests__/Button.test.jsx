import React from 'react';
import renderer from 'react-test-renderer';
import {
  DocumentIcon,
} from '@heroicons/react/outline';
import { Button } from '../Button';

it('should render with default', () => {
  const component = renderer.create(
    <Button />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with children Text', () => {
  const component = renderer.create(
    <Button>Text</Button>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with children DocumentIcon', () => {
  const component = renderer.create(
    <Button><DocumentIcon className="w-5 h-5" /></Button>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with outline', () => {
  const component = renderer.create(
    <Button
      outline
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render without outline', () => {
  const component = renderer.create(
    <Button
      outline={false}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with variant primary', () => {
  const component = renderer.create(
    <Button
      variant="primary"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with variant secondary', () => {
  const component = renderer.create(
    <Button
      variant="secondary"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with variant tertiary', () => {
  const component = renderer.create(
    <Button
      variant="tertiary"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with size base', () => {
  const component = renderer.create(
    <Button
      size="base"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with size sm', () => {
  const component = renderer.create(
    <Button
      size="sm"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with size md', () => {
  const component = renderer.create(
    <Button
      size="md"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with size lg', () => {
  const component = renderer.create(
    <Button
      size="lg"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with size icon', () => {
  const component = renderer.create(
    <Button
      size="icon"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with rounded none', () => {
  const component = renderer.create(
    <Button
      rounded="none"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with rounded sm', () => {
  const component = renderer.create(
    <Button
      rounded="sm"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with rounded md', () => {
  const component = renderer.create(
    <Button
      rounded="md"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with rounded lg', () => {
  const component = renderer.create(
    <Button
      rounded="lg"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with rounded full', () => {
  const component = renderer.create(
    <Button
      rounded="full"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with type button', () => {
  const component = renderer.create(
    <Button
      type="button"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with type submit', () => {
  const component = renderer.create(
    <Button
      type="submit"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with type reset', () => {
  const component = renderer.create(
    <Button
      type="reset"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with disabled true', () => {
  const component = renderer.create(
    <Button
      disabled
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with disabled false', () => {
  const component = renderer.create(
    <Button
      disabled={false}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with tabIndex 0', () => {
  const component = renderer.create(
    <Button
      tabIndex="0"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with disabled false', () => {
  const component = renderer.create(
    <Button
      onClick={() => {}}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with className which makes the colour of the text red', () => {
  const component = renderer.create(
    <Button
      className="text-red-50"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render with arialabel', () => {
  const component = renderer.create(
    <Button
      ariaLabel="this is a button"
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
