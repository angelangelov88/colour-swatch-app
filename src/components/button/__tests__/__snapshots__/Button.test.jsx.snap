// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with arialabel 1`] = `
<button
  aria-label="this is a button"
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with children DocumentIcon 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
>
  <svg
    aria-hidden="true"
    className="w-5 h-5"
    fill="none"
    stroke="currentColor"
    strokeWidth={2}
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7 21h10a2 2 0 002-2V9.414a1 1 0 00-.293-.707l-5.414-5.414A1 1 0 0012.586 3H7a2 2 0 00-2 2v14a2 2 0 002 2z"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
  </svg>
</button>
`;

exports[`should render with children Text 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
>
  Text
</button>
`;

exports[`should render with className which makes the colour of the text red 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg text-red-50"
  disabled={false}
  type="button"
/>
`;

exports[`should render with default 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with disabled false 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with disabled false 2`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  onClick={[Function]}
  type="button"
/>
`;

exports[`should render with disabled true 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg disabled:cursor-default"
  disabled={true}
  type="button"
/>
`;

exports[`should render with outline 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow border border-primary-dark disabled:border-primary-20 text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with rounded full 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-full p-2"
  disabled={false}
  type="button"
/>
`;

exports[`should render with rounded lg 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with rounded md 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-md"
  disabled={false}
  type="button"
/>
`;

exports[`should render with rounded none 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-none"
  disabled={false}
  type="button"
/>
`;

exports[`should render with rounded sm 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-sm"
  disabled={false}
  type="button"
/>
`;

exports[`should render with size base 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with size icon 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow p-2 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with size lg 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-xl py-3 px-6 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with size md 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-lg py-2.5 px-5 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with size sm 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with tabIndex 0 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  tabIndex="0"
  type="button"
/>
`;

exports[`should render with type button 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with type reset 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="reset"
/>
`;

exports[`should render with type submit 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="submit"
/>
`;

exports[`should render with variant primary 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with variant secondary 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-gray-40 active:text-gray-100 text-gray-90 hover:bg-gray-10 active:bg-gray-20 disabled:bg-gray-10 bg-white shadow border-gray-50 text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render with variant tertiary 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-30 hover:text-primary-blue text-primary-light hover:bg-primary-10 active:bg-primary-20 disabled:bg-white bg-white disabled:border-primary-20 text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;

exports[`should render without outline 1`] = `
<button
  className="font-ibm inline-flex items-center font-medium capitalize whitespace-nowrap justify-center focus:outline-none focus:ring-4 focus:ring-offset-0 focus:ring-primary-30 disabled:text-primary-20 text-white hover:bg-primary-blue active:bg-primary-dark disabled:bg-primary-10 bg-primary-light shadow text-sm py-2 px-4 rounded-lg"
  disabled={false}
  type="button"
/>
`;
